File: php/test/CardEntityTest.php
<?php
declare(strict_types=1);

// Card entity test

require_once __DIR__ . '/../trello_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class CardEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = TrelloSDK::test(null, null);
        $ent = $testsdk->Card(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "card" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = TrelloSDK::test($seed, null);
        $seen = iterator_to_array($base->Card(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = TrelloConfig::shared_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = TrelloSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->Card(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = card_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "list", "update", "load", "remove"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "card." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set TRELLO_TEST_CARD_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $card_ref01_ent = $client->Card(null);
        $card_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.card"), "card_ref01"));
        $card_ref01_data["action_id"] = $setup["idmap"]["action01"];
        $card_ref01_data["board_id"] = $setup["idmap"]["board01"];
        $card_ref01_data["list_id"] = $setup["idmap"]["list01"];
        $card_ref01_data["member_id"] = $setup["idmap"]["member01"];

        $card_ref01_data_result = $card_ref01_ent->create($card_ref01_data, null);
        $card_ref01_data = Helpers::to_map(is_object($card_ref01_data_result) && method_exists($card_ref01_data_result, 'data_get') ? $card_ref01_data_result->data_get() : $card_ref01_data_result);
        $this->assertNotNull($card_ref01_data);
        $this->assertNotNull($card_ref01_data["id"]);

        // LIST
        $card_ref01_match = [
            "list_id" => $setup["idmap"]["list01"],
        ];

        $card_ref01_list_result = $card_ref01_ent->list($card_ref01_match, null);
        $this->assertIsArray($card_ref01_list_result);

        $found_item = sdk_select(
            Runner::entity_list_to_data($card_ref01_list_result),
            ["id" => $card_ref01_data["id"]]);
        $this->assertNotEmpty($found_item);

        // UPDATE
        $card_ref01_data_up0_up = [
            "id" => $card_ref01_data["id"],
        ];

        $card_ref01_markdef_up0_name = "address";
        $card_ref01_markdef_up0_value = "Mark01-card_ref01_" . $setup["now"];
        $card_ref01_data_up0_up[$card_ref01_markdef_up0_name] = $card_ref01_markdef_up0_value;

        $card_ref01_resdata_up0_result = $card_ref01_ent->update($card_ref01_data_up0_up, null);
        $card_ref01_resdata_up0 = Helpers::to_map(is_object($card_ref01_resdata_up0_result) && method_exists($card_ref01_resdata_up0_result, 'data_get') ? $card_ref01_resdata_up0_result->data_get() : $card_ref01_resdata_up0_result);
        $this->assertNotNull($card_ref01_resdata_up0);
        $this->assertEquals($card_ref01_resdata_up0["id"], $card_ref01_data_up0_up["id"]);
        $this->assertEquals($card_ref01_resdata_up0[$card_ref01_markdef_up0_name], $card_ref01_markdef_up0_value);

        // LOAD
        $card_ref01_match_dt0 = [
            "id" => $card_ref01_data["id"],
        ];
        $card_ref01_data_dt0_loaded = $card_ref01_ent->load($card_ref01_match_dt0, null);
        $card_ref01_data_dt0_load_result = Helpers::to_map(is_object($card_ref01_data_dt0_loaded) && method_exists($card_ref01_data_dt0_loaded, 'data_get') ? $card_ref01_data_dt0_loaded->data_get() : $card_ref01_data_dt0_loaded);
        $this->assertNotNull($card_ref01_data_dt0_load_result);
        $this->assertEquals($card_ref01_data_dt0_load_result["id"], $card_ref01_data["id"]);

        // REMOVE
        $card_ref01_match_rm0 = [
            "id" => $card_ref01_data["id"],
        ];
        $card_ref01_ent->remove($card_ref01_match_rm0, null);

        // LIST
        $card_ref01_match_rt0 = [
            "list_id" => $setup["idmap"]["list01"],
        ];

        $card_ref01_list_rt0_result = $card_ref01_ent->list($card_ref01_match_rt0, null);
        $this->assertIsArray($card_ref01_list_rt0_result);

        $not_found_item = sdk_select(
            Runner::entity_list_to_data($card_ref01_list_rt0_result),
            ["id" => $card_ref01_data["id"]]);
        $this->assertEmpty($not_found_item);

    }
}

function card_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/card/CardTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = TrelloSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["card01", "card02", "card03", "action01", "action02", "action03", "board01", "board02", "board03", "checklist01", "checklist02", "checklist03", "list01", "list02", "list03", "member01", "member02", "member03", "notification01", "notification02", "notification03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("TRELLO_TEST_CARD_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "TRELLO_TEST_CARD_ENTID" => $idmap,
        "TRELLO_TEST_LIVE" => "FALSE",
        "TRELLO_TEST_EXPLAIN" => "FALSE",
        "TRELLO_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["TRELLO_TEST_CARD_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["TRELLO_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["TRELLO_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new TrelloSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["TRELLO_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["TRELLO_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: go/test/card_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/trello-sdk/go"
	"github.com/voxgig-sdk/trello-sdk/go/core"

	vs "github.com/voxgig-sdk/trello-sdk/go/utility/struct"
)

func TestCardEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Card(nil)
		if ent == nil {
			t.Fatal("expected non-nil CardEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"card": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Card(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Card(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := cardBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "card." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set TRELLO_TEST_CARD_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		cardRef01Ent := client.Card(nil)
		cardRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "card"}), "card_ref01"))
		cardRef01Data["action_id"] = setup.idmap["action01"]
		cardRef01Data["board_id"] = setup.idmap["board01"]
		cardRef01Data["list_id"] = setup.idmap["list01"]
		cardRef01Data["member_id"] = setup.idmap["member01"]

		cardRef01DataResult, err := cardRef01Ent.Create(cardRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		cardRef01Data = core.ToMapAny(entityData(cardRef01DataResult))
		if cardRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if cardRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		cardRef01Match := map[string]any{
			"list_id": setup.idmap["list01"],
		}

		cardRef01ListResult, err := cardRef01Ent.List(cardRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		cardRef01List, cardRef01ListOk := cardRef01ListResult.([]any)
		if !cardRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", cardRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(cardRef01List), map[string]any{"id": cardRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		cardRef01DataUp0Up := map[string]any{
			"id": cardRef01Data["id"],
		}

		cardRef01MarkdefUp0Name := "address"
		cardRef01MarkdefUp0Value := fmt.Sprintf("Mark01-card_ref01_%d", setup.now)
		cardRef01DataUp0Up[cardRef01MarkdefUp0Name] = cardRef01MarkdefUp0Value

		cardRef01ResdataUp0Result, err := cardRef01Ent.Update(cardRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		cardRef01ResdataUp0 := core.ToMapAny(entityData(cardRef01ResdataUp0Result))
		if cardRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if cardRef01ResdataUp0["id"] != cardRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if cardRef01ResdataUp0[cardRef01MarkdefUp0Name] != cardRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", cardRef01MarkdefUp0Name, cardRef01ResdataUp0[cardRef01MarkdefUp0Name])
		}

		// LOAD
		cardRef01MatchDt0 := map[string]any{
			"id": cardRef01Data["id"],
		}
		cardRef01DataDt0Loaded, err := cardRef01Ent.Load(cardRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		cardRef01DataDt0LoadResult := core.ToMapAny(entityData(cardRef01DataDt0Loaded))
		if cardRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if cardRef01DataDt0LoadResult["id"] != cardRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		cardRef01MatchRm0 := map[string]any{
			"id": cardRef01Data["id"],
		}
		_, err = cardRef01Ent.Remove(cardRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		cardRef01MatchRt0 := map[string]any{
			"list_id": setup.idmap["list01"],
		}

		cardRef01ListRt0Result, err := cardRef01Ent.List(cardRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		cardRef01ListRt0, cardRef01ListRt0Ok := cardRef01ListRt0Result.([]any)
		if !cardRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", cardRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(cardRef01ListRt0), map[string]any{"id": cardRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func cardBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "card", "CardTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read card test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse card test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"card01", "card02", "card03", "action01", "action02", "action03", "board01", "board02", "board03", "checklist01", "checklist02", "checklist03", "list01", "list02", "list03", "member01", "member02", "member03", "notification01", "notification02", "notification03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("TRELLO_TEST_CARD_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"TRELLO_TEST_CARD_ENTID": idmap,
		"TRELLO_TEST_LIVE":      "FALSE",
		"TRELLO_TEST_EXPLAIN":   "FALSE",
		"TRELLO_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["TRELLO_TEST_CARD_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["TRELLO_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["TRELLO_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewTrelloSDK(core.ToMapAny(mergedOpts))
	}

	live := env["TRELLO_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["TRELLO_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
